### PR TITLE
fix masked_select for discontiguous outputs

### LINKED ISSUE #41473: 'torch.masked_select' behaves differently depending on the device of inputs.
## 🐛 Bug

Outputs from torch.masked_select are different when input tensors' device is cpu and when it's cuda.

## To Reproduce

1. candidate = torch.tensor([[[-1, -2], [0, 1], [2, 3], [4, 5], [6, 7], [8, 9]],
                              [[-11, -12], [10, 11], [12, 13], [14, 15], [16, 17], [18, 19]],
                              [[-101, -102], [100, 101], [102, 103], [104, 105], [106, 107], [108, 109]]])
1. mask = torch.tensor([[0, 1, 1, 1, 1, 0],
                              [0, 1, 1, 0, 1, 1],
                              [0, 1, 1, 0, 1, 1]], dtype=torch.bool)
1. result_cpu = torch.masked_select(candidate.permute(2, 0, 1), mask)
1. result_gpu = torch.masked_select(candidate.to('cuda').permute(2, 0, 1), mask.to('cuda'))
1. assert not (result_cpu != result_gpu.cpu()).any()

## Expected behavior

Outputs (result_cpu and result_gpu) should be identical.
Gettting : 
result_cpu = tensor([  0, 1, 2, 3, 4, 5, 6, 7, 10, 11, 12, 13, 16, 17, 18, 19, 100, 101, 102, 103, 106, 107, 108, 109])
result_gpu = tensor([ 0, 2, 4, 6, 10, 12, 16, 18, 100, 102, 106, 108, 1, 3, 5, 7, 11, 13, 17, 19, 101, 103, 107, 109], device='cuda:0')

## Environment
PyTorch version: 1.7.0a0+c71ec1c
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609
CMake version: version 3.14.0

Python version: 3.8
Is CUDA available: Yes
CUDA runtime version: 10.1.105
GPU models and configuration: GPU 0: Tesla K40m
Nvidia driver version: 440.33.01
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.1.3

Versions of relevant libraries:
[pip3] numpy==1.18.5
[pip3] torch==1.7.0a0+c71ec1c
[conda] blas                      1.0                         mkl    defaults
[conda] magma-cuda101             2.5.2                         1    pytorch
[conda] mkl                       2020.1                      217    defaults
[conda] mkl-include               2020.1                      217    defaults
[conda] mkl-service               2.3.0            py38he904b0f_0    defaults
[conda] mkl_fft                   1.1.0            py38h23d657b_0    defaults
[conda] mkl_random                1.1.1            py38h0573a6f_0    defaults
[conda] numpy                     1.18.5           py38ha1c710e_0    defaults
[conda] numpy-base                1.18.5           py38hde5b4d6_0    defaults
[conda] torch                     1.7.0a0+c71ec1c          pypi_0    pypi

cc @ezyang @gchanan @zou3519

### PR DESCRIPTION
**Pull request by ngimel**
This fixes #41473 for discontiguous input, mask and out. Tests to follow. Reverting #33269 is not a great solution because I'm told masked_select was needed for printing complex tensors. 
cc @gchanan , @zou3519, @ezyang

### CODE REVIEW

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp [hunk 1]
```diff
@@ -715,11 +716,11 @@ static Tensor & masked_select_out_impl_cpu(Tensor & result, const Tensor & self,
       .check_all_same_dtype(false)
       .resize_outputs(false)
       .add_output(result_strided)
-      .add_input(_self)
-      .add_input(_mask)
+      .add_input(_self.contiguous())
+      .add_input(_mask.contiguous())
```
Comment (malfet) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp:720]: This would be a no-op for contiguous tensors, but would create a copy for non-contiguous ones, right?

**Comment by gchanan** [comment]
why don't we have tests for printing complex tensors?  The tests for https://github.com/pytorch/pytorch/pull/41828 pass.

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp (hunk 1 above)
Comment (malfet) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:720]: And why the same fix does not need to be applied to TensorConfig() created on line 737 for parallel kernel?

**Comment by malfet** [comment]
> why don't we have tests for printing complex tensors? The tests for #41828 pass.
We do, see
https://github.com/pytorch/pytorch/blob/fced54aa679afeac0159ca75b40d69d79f2e47fc/test/test_torch.py#L3911-L3921
And for strided ones
https://github.com/pytorch/pytorch/blob/fced54aa679afeac0159ca75b40d69d79f2e47fc/test/test_torch.py#L4052-L4076

**Comment by ngimel** [comment]
Ok, maybe we don't need masked_select for complex printing after all, I just remember @anjali411 wanted original PR for something complex-replated.

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp (hunk 1 above)
Comment (ngimel) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:720]: According to @zou3519 only serial kernel is affected, and I didn't figure out yesterday how to fix the kernel itself. I'll take a look now.

**Comment by gchanan** [comment]
yep, I checked with @anjali411 and she said we split the real and imaginary parts before printing, so we should be good.

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp (hunk 1 above)
Comment (ngimel) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp]: I root-caused why its serial kernel only, so now I'm sending discontiguous inputs to parallel kernel that can handle them without making additional copies.

**Comment by ezyang** [comment]
Thanks for the fix. Is malfet doing the full review? Add me if you want me to review too.

**Comment by ngimel** [comment]
@malfet it's ready for review now. Unfortunately it's harder to review, as after revert I have to bring original porting PR back, but now it has a test that explicitly tests for discontiguous mask/value/out, so if that passes, should be good to go. 
Unlike before, broadcasted mask/value now will also go through parallel kernel (because they are not contiguous), so the perf would be somewhat worse for those cases. Before broadcasted cases could use serial kernel, because TensorIterator normally does not reorder dimensions in this case. We don't have an API to query TensorIterator if dimensions are reordered, and I did not think it makes sense to bring it just for this case.

**Review by mruberry — APPROVED** [review]
Awesome fix to a couple silent errors!
